rpcs3/Emu/Cell/Modules/cellAtracXdec.cpp

```diff
@@ -0,0 +1,884 @@
+		{
+			skip_getting_command = true;
+
+			// Block savestate creation during callbacks
+			std::unique_lock savestate_lock{g_fxo->get<hle_locks_t>(), std::try_to_lock};
+
+			if (!savestate_lock.owns_lock())
+			{
+				ppu.state += cpu_flag::again;
+				return;
+			}
+
+			skip_getting_command = false;
+
+			// Doesn't do anything else
+			notify_seq_done.cbFunc(ppu, notify_seq_done.cbArg);
+			break;
+		}
+		case AtracXdecCmdType::decode_au:
+		{
+			skip_getting_command = true;
+
+			ensure(!!cmd.au_start_addr); // Not checked on LLE
+
+			cellAtracXdec.trace("Waiting for output to be consumed...");
+
+			lv2_obj::sleep(ppu);
+			std::unique_lock output_mutex_lock{output_mutex};
+
+			while (output_locked && !ppu.is_stopped())
+			{
+				lv2_obj::sleep(ppu);
+				output_consumed.wait(output_mutex, 20000);
+			}
+
+			if (ppu.is_stopped())
+			{
+				ppu.state += cpu_flag::again;
+				return;
+			}
+
+			if (!run_thread)
+			{
+				return;
+			}
+
+			cellAtracXdec.trace("Output consumed");
+
+			u32 error = CELL_OK;
+
+			// Only the first valid ATS header after starting a sequence is parsed. It is ignored on all subsequent access units
+			if (first_decode && atracx_param.au_includes_ats_hdr_flg == CELL_ADEC_ATRACX_ATS_HDR_INC)
+			{
+				// Block savestate creation during callbacks
+				std::unique_lock savestate_lock{g_fxo->get<hle_locks_t>(), std::try_to_lock};
+
+				if (!savestate_lock.owns_lock())
+				{
+					ppu.state += cpu_flag::again;
+					return;
+				}
+
+				if (error = decoder.parse_ats_header(cmd.au_start_addr); error != CELL_OK)
+				{
+					notify_error.cbFunc(ppu, error, notify_error.cbArg);
+				}
+				else if (decoder.init_decode(atracx_param.bw_pcm, atracx_param.nch_out) != CELL_OK)
+				{
+					notify_error.cbFunc(ppu, CELL_ADEC_ERROR_ATX_FATAL, notify_error.cbArg);
+				}
+			}
+
+			// LLE does not initialize the output address if parsing the ATS header fails
+			vm::ptr<void> output = vm::null;
+			u32 decoded_samples_num = 0;
+
+			if (error != CELL_ADEC_ERROR_ATX_ATSHDR)
+			{
+				// The LLE SPU thread would crash if you attempt to decode without a valid configuration
+				ensure(decoder.config_is_set, "Attempted to decode with invalid configuration");
+
+				output.set(work_mem.addr() + atracXdecGetSpursMemSize(decoder.nch_in));
+
+				const auto au_start_addr = atracx_param.au_includes_ats_hdr_flg == CELL_ADEC_ATRACX_ATS_HDR_INC ? cmd.au_start_addr.get_ptr() + sizeof(AtracXdecAtsHeader) : cmd.au_start_addr.get_ptr();
+				std::memcpy(work_mem.get_ptr(), au_start_addr, decoder.nbytes);
+
+				if (int err = avcodec_send_packet(decoder.ctx, decoder.packet); err)
+				{
+					cellAtracXdec.error("avcodec_send_packet() failed (err=0x%x='%s')", err, utils::av_error_to_string(err));
+					error = CELL_ADEC_ERROR_ATX_NON_FATAL; // Not accurate, FFmpeg doesn't provide detailed errors like LLE
+				}
+
+				if (int err = avcodec_receive_frame(decoder.ctx, decoder.frame); err != 0 && err != AVERROR(EAGAIN))
```

Comment:
hmm. I don't think I explained this properly:
- You can't call `avcodec_send_packet` again while the decoder still holds frames. It will then also return EAGAIN to signify that you have to call `avcodec_receive_frame` again (currently you would log an error above in this case and drop the packet).
- Respectively, if `avcodec_receive_frame` returns EAGAIN, the frame isn't ready and you have to wait for the next packet and call `avcodec_send_packet` and then `avcodec_receive_frame` again etc. before you can process the frame below.
- The decoder may need n packets to produce x frames. It's not necessarily a 1 to 1 conversion.

So this basically means:
- Keep calling `avcodec_receive_frame` while it returns success until you extracted all the valid frames and it returns EAGAIN (or AVERROR_EOF)
- When `avcodec_receive_frame` returns EAGAIN (or AVERROR_EOF): No more frames are ready (or the frame needs more input), wait for next `avcodec_send_packet`
- Make sure you don't call `avcodec_send_packet` while the decoder still holds frames

It's possible that this will practically never happen with these codecs, but since this is how ffmpeg works (and to be on the safe side) you have to design your code to be able to handle more than one decoded frame per decode_au call.

Reply:
I looked into this, and I'm pretty sure that if the AVPacket only contains a single encoded frame, then the decoder will only output a single AVFrame. Apparently, some of FFmpeg's demuxers can output AVPackets containing more than one encoded frame. Since I'm truncating the user input to exactly one frame,
https://github.com/RPCS3/rpcs3/blob/5829b4e2ad4e819f431424d1db7ea0dac463b4a3/rpcs3/Emu/Cell/Modules/cellAtracXdec.cpp#L409-L412
calling `avcodec_receive_frame` multiple times should never be necessary for this implementation.

Additionally, there appears to be a flag that indicates if a decoder can output multiple frames per packet:
https://github.com/RPCS3/ffmpeg-core/blob/44845469c918c07d8470df2330711ed28174c468/include/libavcodec/codec.h#L89-L100
This flag is not set for the atrac3plus decoder.

Either way, I cannot properly implement this because there is only enough space for exactly one frame in the output buffer. Since the user can get a pointer to the buffer with `cellAdecGetPcmItem`, I cannot use host memory without breaking that.

Reply:
Well then you should probably remove the eagain from the receive again